auto_dev/protocols/primitives.py

```diff
@@ -0,0 +1,233 @@
+min_uint64 = 0
+max_uint64 = (1 << 64) - 1
+
+min_float32 = struct.unpack("f", struct.pack("I", 0xFF7FFFFF))[0]
+max_float32 = struct.unpack("f", struct.pack("I", 0x7F7FFFFF))[0]
+min_float64 = struct.unpack("d", struct.pack("Q", 0xFFEFFFFFFFFFFFFF))[0]
+max_float64 = struct.unpack("d", struct.pack("Q", 0x7FEFFFFFFFFFFFFF))[0]
+
+
+def to_float32(value: float) -> float:
+    """Pack the value as a 32-bit float then unpack it."""
+    return struct.unpack("f", struct.pack("f", value))[0]
+
+
+class BaseConstrainedFloat(float, ABC):
+    """Base class for constrained float types."""
+
+    @classmethod
+    @abstractmethod
+    def min(cls) -> float:
+        msg = f"{cls.__name__}.min() is not implemented."
+        raise NotImplementedError(msg)
+
+    @classmethod
+    @abstractmethod
+    def max(cls) -> float:
+        msg = f"{cls.__name__}.max() is not implemented."
+        raise NotImplementedError(msg)
+
+    def __new__(cls, value: float = 0.0, *args, **kwargs) -> "BaseConstrainedInt":
+        schema = core_schema.float_schema(strict=True, ge=cls.min(), le=cls.max())
+        validator = SchemaValidator(schema)
+        validated_value = validator.validate_python(value)
+        return super().__new__(cls, validated_value)
+
+    @classmethod
+    def __get_pydantic_core_schema__(cls, source, handler):
+        schema = core_schema.float_schema(strict=True, ge=cls.min(), le=cls.max())
+        return core_schema.no_info_wrap_validator_function(cls, schema)
+
+
+class BaseConstrainedInt(int, ABC):
+    """Base class for constrained integer types."""
+
+    @classmethod
+    @abstractmethod
+    def min(cls) -> int:
+        msg = f"{cls.__name__}.min() is not implemented."
+        raise NotImplementedError(msg)
+
+    @classmethod
+    @abstractmethod
+    def max(cls) -> int:
+        msg = f"{cls.__name__}.max() is not implemented."
+        raise NotImplementedError(msg)
+
+    def __new__(cls, value: int = 0, *args, **kwargs) -> "BaseConstrainedInt":
+        schema = core_schema.int_schema(strict=True, ge=cls.min(), le=cls.max())
+        validator = SchemaValidator(schema)
+        validated_value = validator.validate_python(value)
+        return super().__new__(cls, validated_value)
+
+    @classmethod
+    def __get_pydantic_core_schema__(cls, source, handler):
+        schema = core_schema.int_schema(strict=True, ge=cls.min(), le=cls.max())
+        return core_schema.no_info_wrap_validator_function(cls, schema)
+
+
+class Double(BaseConstrainedFloat):
+    @classmethod
+    def min(cls):
+        return min_float64
+
+    @classmethod
+    def max(cls):
+        return max_float64
+
+
+class Float(BaseConstrainedFloat):
+    @classmethod
+    def min(cls):
+        return min_float32
+
+    @classmethod
+    def max(cls):
+        return max_float32
+
+    def __new__(cls, value: float = 0.0, *args, **kwargs) -> "Float":
+        return super().__new__(cls, to_float32(float(value)))
+
+
+class Int32(BaseConstrainedInt):
```

Comment:
we need this,not to constrain the data per se (pydantic can deal with that via i.e. `conint`), but to encode to the correct proto type; the information from i.e. conint does not carry over to the python `int` generated, so in many cases you cannot discern which type of proto integer it was  / needs to adhere to